tensorflow/core/grappler/optimizers/data/filter_with_random_uniform_fusion.cc

```diff
@@ -0,0 +1,206 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/grappler/optimizers/data/filter_with_random_uniform_fusion.h"
+
+#include "tensorflow/core/framework/attr_value.pb.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/grappler/clusters/cluster.h"
+#include "tensorflow/core/grappler/graph_view.h"
+#include "tensorflow/core/grappler/grappler_item.h"
+#include "tensorflow/core/grappler/op_types.h"
+#include "tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.h"
+#include "tensorflow/core/grappler/optimizers/data/graph_utils.h"
+#include "tensorflow/core/grappler/utils.h"
+#include "tensorflow/core/platform/protobuf.h"
+
+#include <iostream>
+
+namespace tensorflow {
+namespace grappler {
+
+constexpr char kFusedOpName[] = "SamplingDataset";
+
+NodeDef MakeFusedNode(const NodeDef& filter_node, float rate,
+                      MutableGraphView* graph) {
+  NodeDef fused_node;
+  graph_utils::SetUniqueGraphNodeName("fused_sampling", graph->graph(),
+                                      &fused_node);
+  fused_node.set_op(kFusedOpName);
+
+  // Copy over inputs.
+  for (int i = 0; i < filter_node.input_size(); ++i) {
+    fused_node.add_input(filter_node.input(i));
+  }
+
+  // Required attrs.
+  for (auto key : {"output_shapes", "output_types"}) {
+    graph_utils::CopyAttribute(key, filter_node, &fused_node);
+  }
+
+  // Optional attrs.
+  for (auto key : {"use_inter_op_parallelism", "sloppy"}) {
+    if (gtl::FindOrNull(filter_node.attr(), key)) {
+      graph_utils::CopyAttribute(key, filter_node, &fused_node);
+    }
+  }
+
+  NodeDef* tmp = graph_utils::AddScalarConstNode<float>(rate, graph);
+  fused_node.add_input(tmp->name());
+
+  return fused_node;
+}
+
+const NodeDef* FunctionFindNodeDef(const FunctionDef& function, const string op,
+                                   const string func, const string match) {
+  for (const NodeDef& func_node : function.node_def()) {
+    if (func_node.op() != op) {
+      continue;
+    }
+    if (func_node.name() + match != func) {
+      continue;
+    }
+    return &func_node;
+  }
+  return NULL;
+}
+
+// This optimization fuse one of the following two forms of
+// filter + random_uniform predication into a single data sampling operation:
+// fuse:
+//   filter
+//   |
+//   + predication: less [0]
+//                  |
+//                  + random_uniform [1]
+//                  |
+//                  + rate
+// or:
+//   filter
+//   |
+//   + predication: less
+//                  |
+//                  + random_uniform[]
+//                  |
+//                  + rate
+// into:
+//   sampling(rate)
+Status FilterWithRandomUniformFusion::OptimizeAndCollectStats(Cluster* cluster,
+                                                  const GrapplerItem& item,
```

Comment:
the indentation here looks wrong

Reply:
Fixed in https://github.com/tensorflow/tensorflow/pull/25384/commits/fa11069c3e85ffbcae19994bdba09c30eee0a6de#diff-ccdaa4724ea485e41d03d236b6ee46d0R129